File: kotlin/src/main/kotlin/io/github/piszmog/aoc/Day2.kt
package io.github.piszmog.aoc

import java.time.Instant

fun main(args: Array<String>) {
    val start = Instant.now()

    val parser = getCSVParser(args[0], " ")
    val courses = parser.map { Course(Direction.valueOf(it[0].uppercase()), it[1].toInt()) }.toList()

    val part1Solution = day2Part1(courses)
    println("Part 1: $part1Solution")
    val part2Solution = day2Part2(courses)
    println("Part 2: $part2Solution")

    printElapsedTime(start)
}

fun day2Part1(input: List<Course>): Int {
    var horizontalSum = 0
    var depthSum = 0
    input.forEach {
        when (it.direction) {
            Direction.FORWARD -> horizontalSum += it.value
            Direction.DOWN -> depthSum += it.value
            Direction.UP -> depthSum -= it.value
        }
    }
    return horizontalSum * depthSum
}

fun day2Part2(input: List<Course>): Int {
    var horizontalSum = 0
    var depthSum = 0
    var aimSum = 0
    input.forEach {
        when (it.direction) {
            Direction.FORWARD -> {
                horizontalSum += it.value
                depthSum += it.value * aimSum
            }
            Direction.DOWN -> aimSum += it.value
            Direction.UP -> aimSum -= it.value
        }
    }
    return horizontalSum * depthSum
}

data class Course(val direction: Direction, val value: Int)

enum class Direction {
    DOWN, FORWARD, UP
}

File: rust/src/bin/day2.rs
use std::time::Instant;

use serde::Deserialize;

use aoc_2021::shared;

#[derive(Debug, Deserialize)]
struct Course {
    direction: String,
    value: i32,
}

fn main() {
    let start = Instant::now();
    let mut reader = shared::get_csv_reader("../inputs/day2.txt", false, b' ').unwrap();
    let courses: Vec<Course> = reader.deserialize().map(|result| {
        let course: Course = result.unwrap();
        course
    }).collect();

    let part_1_solution = part_1(&courses);
    println!("Part 1: {}", part_1_solution);
    let part_2_solution = part_2(&courses);
    println!("Part 2: {}", part_2_solution);

    shared::print_elapsed_time(start);
}

fn part_1(input: &Vec<Course>) -> i32 {
    let mut horizontal_sum = 0;
    let mut depth_sum = 0;
    input.iter().for_each(|course| {
        match course.direction.as_str() {
            "forward" => horizontal_sum += course.value,
            "down" => depth_sum += course.value,
            "up" => depth_sum -= course.value,
            _ => {}
        }
    });
    horizontal_sum * depth_sum
}

fn part_2(input: &Vec<Course>) -> i32 {
    let mut horizontal_sum = 0;
    let mut depth_sum = 0;
    let mut aim_sum = 0;
    input.iter().for_each(|course| {
        match course.direction.as_str() {
            "forward" => {
                horizontal_sum += course.value;
                depth_sum += course.value * aim_sum;
            }
            "down" => aim_sum += course.value,
            "up" => aim_sum -= course.value,
            _ => {}
        }
    });
    horizontal_sum * depth_sum
}

#[cfg(test)]
mod day2_tests {
    use std::str::FromStr;

    use crate::{Course, part_1, part_2};

    #[test]
    fn test_part_1() {
        let input = get_input();
        let actual = part_1(&input);
        assert_eq!(150, actual);
    }

    #[test]
    fn test_part_2() {
        let input = get_input();
        let actual = part_2(&input);
        assert_eq!(900, actual);
    }

    fn get_input() -> Vec<Course> {
        let raw_input = vec!(
            "forward 5",
            "down 5",
            "forward 8",
            "up 3",
            "down 8",
            "forward 2",
        );
        raw_input.iter()
            .map(|s| s.split(" "))
            .map(|mut val| Course {
                direction: val.next().unwrap().to_string(),
                value: FromStr::from_str(val.next().unwrap()).unwrap(),
            })
            .collect()
    }
}